### PR TITLE
Fixing Python 3.x compatibility for Sets.

### PR DESCRIPTION
**Pull request by jsiirola**
See the following discussion on the forum: https://groups.google.com/d/msg/pyomo-forum/XOf6zwvEbt4/2biHLHRRAwAJ

Two key fixes:
 - Set.pprint() failed for mixed-type Sets in Python3
 - Set() failed to create a clonable model when initialized using dict.keys() in Python3

### CODE REVIEW

File: pyomo/core/base/sets.py [hunk 1]
```diff
@@ -110,6 +110,36 @@ def wrapper_function ( *args, **kwargs ):
     return wrapper_function
 
 
+class _robust_sorter(object):
```
Comment (carldlaird) [inline pyomo/core/base/sets.py]: I understand what this is doing after speaking with John. However, this class needs some more documentation to make it clear to others later.

**Comment by ckaldemeyer** [comment]
> See the following discussion on the forum: https://groups.google.com/d/msg/pyomo-forum/XOf6zwvEbt4/2biHLHRRAwAJ
> 
> Two key fixes:
> 
>     Set.pprint() failed for mixed-type Sets in Python3
>     Set() failed to create a clonable model when initialized using dict.keys() in Python3

Can you already estimate when this PR will be merged and when the next release will be announced?

Depending on this, I would decide to wait for your release or (temporarily) fix all my problem formulations with the hotfix you suggested in the forum.

Best
Cord

**Comment by jsiirola** [comment]
I am hoping that @carldlaird or @whart222 will review the PR in the next day or two and get it merged.  We are actively preparing for the 5.4 release (see http://github.com/Pyomo/pyomo/projects/3), but I hope that that release will be out soon (ideally, within a week or two)

File: pyomo/core/base/sets.py [hunk 2]
```diff
@@ -110,6 +110,72 @@ def wrapper_function ( *args, **kwargs ):
     return wrapper_function
 
 
+class _robust_sort_keys(object):
```
Comment (whart222) [inline pyomo/core/base/sets.py]: Why is this a class?  Why not provide a function, which is a simpler code design?

File: pyomo/core/base/sets.py [hunk 3]
```diff
@@ -110,6 +110,72 @@ def wrapper_function ( *args, **kwargs ):
...
+        return _type.__name__, self._typemap[_type](val)
+
+
+
+def _value_sorter(self, obj):
+    """Utility to sort the values of a Set.
+
+    This returns the values of the Set in a consistent order.  For
+    ordered Sets, simply return the ordered list.  For unordered Sets,
+    first try the standard sorted order, and if that failes (for example
```
Comment (whart222) [inline pyomo/core/base/sets.py]: failes -> fails

**Comment by whart222** [comment]
@carldlaird confirmed that the requested documentation has been added to this PR.

**Comment by ckaldemeyer** [comment]
> I am hoping that @carldlaird or @whart222 will review the PR in the next day or two and get it merged. We are actively preparing for the 5.4 release (see http://github.com/Pyomo/pyomo/projects/3), but I hope that that release will be out soon (ideally, within a week or two)

Thanks for your answer and your effort. I'll then just wait for the next release!

**Pull request merged**